test/Dialect/FIRRTL/grand-central.mlir

```diff
@@ -1123,6 +1123,220 @@ firrtl.circuit "SubmoduleDirectoryBehavior" attributes {
+// CHECK-NOT:    output_file
+//
+// CHECK:      firrtl.module
+
+// -----
+
+firrtl.circuit "DirectoryBehaviorWithoutDUT" attributes {
+  annotations = [
+    {class = "sifive.enterprise.grandcentral.AugmentedBundleType",
+     defName = "Foo",
+     elements = [],
+     id = 0 : i64,
+     name = "View"},
+    {class = "sifive.enterprise.grandcentral.ExtractGrandCentralAnnotation",
+     directory = "gct-dir",
+     filename = "bindings.sv"}]} {
+
+  // Each of these modules is instantiated in a different location.  A leading
+  // "E" indicates that this is an external module.  A leading "M" indicates
+  // that this is a module.  The instantiation location is indicated by three
+  // binary bits with an "_" indicating the absence of instantiation:
+  //   1) "T" indicates this is instantiated in the "Top"
+  //   2) "C" indicates this is instantiated in the "Companion"
+  // E.g., "E_C" is an external module instantiated only in the Companion.
+  firrtl.module @MT_() {}
+  firrtl.module @M_C() {}
+  firrtl.module @MTC() {}
+  firrtl.extmodule @ET_() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "ET_.v", text = ""}
+  ]}
+  firrtl.extmodule @E_C() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "E_C.v", text = ""}
+  ]}
+  firrtl.extmodule @ETC() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "ETC.v", text = ""}
+  ]}
+
+  // The Grand Central Companion module.
+  firrtl.module private @Companion() attributes {
+    annotations = [
+      {class = "sifive.enterprise.grandcentral.ViewAnnotation.companion",
+       defName = "Foo",
+       id = 0 : i64,
+       name = "View"}]} {
+
+    firrtl.instance m__c @M_C()
+    firrtl.instance m_dc @MTC()
+
+    firrtl.instance e__c @E_C()
+    firrtl.instance e_dc @ETC()
+  }
+
+  // This is the DUT in the previous example, but is no longer marked as the
+  // DUT.
+  firrtl.module @DirectoryBehaviorWithoutDUT() attributes {
+    annotations = [
+      {class = "sifive.enterprise.grandcentral.ViewAnnotation.parent",
+       id = 0 : i64,
+       name = "view"}
+    ]} {
+    firrtl.instance companion @Companion()
+
+    firrtl.instance m_d_ @MT_()
+    firrtl.instance m_dc @MTC()
+
+    firrtl.instance e_d_ @ET_()
+    firrtl.instance e_dc @ETC()
+  }
+
+}
+
+// Any module instantiated by the Companion, but not instantiated by the DUT is
+// moved to the same directory as the Companion.  I.e., only "*_C" modules
+// should be moved into the "gct-dir".
+//
+// CHECK-LABEL: "DirectoryBehaviorWithoutDUT"
+//
+// CHECK-NOT:    output_file
+// CHECK:      firrtl.module @M_C
+// CHECK-SAME:   output_file = #hw.output_file<"gct-dir{{/|\\\\}}"
+// CHECK-NOT:    output_file
+//
+// CHECK:      firrtl.extmodule @E_C
+// CHECK-SAME:   output_file = #hw.output_file<"gct-dir{{/|\\\\}}">
+// CHECK-NOT:    output_file
+//
+// CHECK:      firrtl.module
```

Comment:
This test is the following diagram. This is testing that "B" gets moved under the companion. This is all of `*_C`:

![image](https://user-images.githubusercontent.com/1018530/199359535-e3e7df9e-6ab2-422c-b4c3-ccf665c8f1c4.png)